File: 2405.1/jf_syn/android/AndroidUnsafeHostnameVerificationChecker/AndroidUnsafeHostnameVerification.java
import android.app.Activity;

import java.security.cert.Certificate;
import java.security.cert.X509Certificate;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLException;
import javax.net.ssl.SSLSession;

public class AndroidUnsafeHostnameVerification extends Activity {
    public void violation() {
        HostnameVerifier verifier = new HostnameVerifier() {
            @Override
            public boolean verify(String hostname, SSLSession session) {
                return true;                                                // @violation
            }
        };
        HttpsURLConnection.setDefaultHostnameVerifier(verifier);
    }

    public void good() {
        HostnameVerifier verifier = new HostnameVerifier() {
            @Override
            public boolean verify(String hostname, SSLSession session) {
                try {
                    Certificate[] certs = session.getPeerCertificates();
                    X509Certificate x509 = (X509Certificate) certs[0];
                    check(new String[]{host}, x509);
                    return true;                                            /* SAFE */
                } catch (SSLException e) {
                    return false;
                }
            }
        };
        HttpsURLConnection.setDefaultHostnameVerifier(verifier);
    }
}

File: 2405.1/kotlinchk/security/AndroidUnsafeHostnameVerificationChecker/AndroidUnsafeHostnameVerification.kt
package testcase.rule.security.AndroidUnsafeHostnameVerificationChecker

import java.security.cert.Certificate
import java.security.cert.X509Certificate
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.HttpsURLConnection
import javax.net.ssl.SSLException
import javax.net.ssl.SSLSession

object TestCase {
	fun violation1() {
		val verifier = HostnameVerifier { hostname, session -> true }				// @violation
		HttpsURLConnection.setDefaultHostnameVerifier(verifier)
	}

	fun violation2() {
		val verifier = object : HostnameVerifier {
			override fun verify(hostname: String, session: SSLSession): Boolean {	// @violation
				return true
			}
		}
		HttpsURLConnection.setDefaultHostnameVerifier(verifier)
	}

	fun good1() {
		val verifier = HostnameVerifier { hostname, session ->						// good
			try {
				val certs: Array<Certificate> = session.peerCertificates
				val x509: X509Certificate = certs[0] as X509Certificate
				check(arrayOf<String>(host), x509)
				true
			} catch (e: SSLException) {
				false
			}
		}
		HttpsURLConnection.setDefaultHostnameVerifier(verifier)
	}

	fun good2() {
		val verifier = object : HostnameVerifier {
			override fun verify(hostname: String, session: SSLSession): Boolean {	// good
				return try {
					val certs: Array<Certificate> = session.peerCertificates
					val x509: X509Certificate = certs[0] as X509Certificate
					check(arrayOf<String>(host), x509)
					true
				} catch (e: SSLException) {
					false
				}
			}
		}
		HttpsURLConnection.setDefaultHostnameVerifier(verifier)
	}

	fun good3() {
		val verifier = object : HostnameVerifier {
			override fun verify(hostname: String, session: SSLSession): Boolean {	// good
				try {
					val certs: Array<Certificate> = session.peerCertificates
					val x509: X509Certificate = certs[0] as X509Certificate
					check(arrayOf<String>(host), x509)
					return true
				} catch (e: SSLException) {
					return false
				}
			}
		}
		HttpsURLConnection.setDefaultHostnameVerifier(verifier)
	}
}